polygeist/tools/cgeist/Lib/CodeGenTypes.cc

```diff
@@ -0,0 +1,642 @@
+//===--- CodeGenTypes.cc - Type translation for MLIR CodeGen -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This is the code that handles AST -> MLIR type lowering.
+//
+//===----------------------------------------------------------------------===//
+
+#include "CodeGenTypes.h"
+#include "TypeUtils.h"
+#include "utils.h"
+
+#include "clang/AST/ASTContext.h"
+#include "clang/CodeGen/CGFunctionInfo.h"
+
+#include "clang/../../lib/CodeGen/CodeGenModule.h"
+#include "clang/../../lib/CodeGen/TargetInfo.h"
+
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOpsTypes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Target/LLVMIR/TypeFromLLVM.h"
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "cgeist"
+
+using namespace clang;
+using namespace mlir;
+using namespace llvm;
+
+static cl::opt<bool>
+    memRefFullRank("memref-fullrank", cl::init(false),
+                   cl::desc("Get the full rank of the memref."));
+
+static cl::opt<bool> memRefABI("memref-abi", cl::init(true),
+                               cl::desc("Use memrefs when possible"));
+
+/******************************************************************************/
+/*            Flags affecting code generation of function types.              */
+/******************************************************************************/
+
+// Note: cgeist does not allow flattening struct function parameters. Need to
+// revisit.
+constexpr bool AllowStructFlattening = false;
+
+// Note: cgeist does not allow returning a struct via the parameter list. Need
+// to revisit.
+constexpr bool AllowSRet = false;
+
+// Note: cgesit does not allow returning 'inalloca'. Need to revisit.
+constexpr bool AllowInAllocaRet = false;
```

Comment:
Can we add these flags in the non-nfc pr?

Reply:
done